acarsdec next? :-)

f00b4r0 opened the pull request:
Hi,

This large PR proposes what could be acarsdec 4.x

It'll be too long to document each commit separately (that's what the commit log is for anyway, I tried to use meaningful commit messages) so I'll focus on main changes:

This PR includes #116, #107, #106 and bits of #111, in more or less reworked form. It fixes #112, #76 and presumably #100 (through inclusion of #107).

The bulk of this PR consists of:
- a large code cleanup and refactor, reducing the number of duplicated (sometimes with slight unnecessary variations) code, deleting dead code and removing unnecessary bandaids; hence deleting more lines than are added (despite adding several header files and comment blocks). The net result is improved memory footprint and CPU usage.
- improving error checking and reporting
- improving signal handling and exit cleanup
- improving build system: all optional libraries are now autodetected (including libasound for ALSA support), the system-wide libcJSON is used instead of embedding a local copy and it builds fine on Linux and macOS (cannot test Windows but I don't expect changes there from previous situation)
- **enabling an unlimited number of ACARS channels to be decoded simultaneously** (within available bandwidth and CPU power). This also means that acarsdec now uses just as much memory it needs for the number of channels *actually* decoded, no more, no less.
- **allowing all supported inputs to be built in the same executable** (no more "cmake -Drtl=ON") which, together with the removal of the unconditional `-march=native` compilation flag (which can be added back on cmake invocation, as detailed in README), should make acarsdec packageable by distributions if necessary
- **allowing all supported output types to be used concurrently and repeatedly**, which means it's now possible to display full text will sending JSON (or anything else) via UDP to *several* feed collectors *and* updating an MQTT broker **at the same time**
- **improving performance-critical common code** which translates into e.g. the CPU usage of the SoapySDR backend being all but **divided by 4**, so that the overhead of using Soapy instead of native backend is now quite negligible
- implementing multichannel raw audio input, at multiple sample rates (partly addresses #109, supersedes #110).
- and more 😁

Oh yes and now there's also a bit of **statistics collection and reporting** available via a simple StatsD-compatible interface. It's loosely based on how dumpvdl2 reports its own statistics. This helps to e.g. keep track of the performance of the SDR dongle, or to find which ACARS channels are the busiest.

The code reorganization (building all supported input backends together) and the ability to use multiple output combinations prompted me to change quite drastically how command line options are processed. This means in particular that from this point forward, acarsdec is **no longer compatible** with previous command line parameters (another reason for increasing the version major number, if all the above changes weren't enough already 😉):
- inputs now use unambiguous long options (e.g. `--rtlsdr` instead of `-r` or `--airspy` instead of `-s`, loosely based on dumpvdl2), yet their argument remains **unchanged**: migration is easy.
- outputs are defined using `--output FORMAT:DESTINATION:PARAMETERS` allowing mix and match of output formats and destinations (e.g. it's now possible to log planeplotter format to file, or send full text via UDP). This is detailed in the README and the supported formats and destinations can be listed via `acarsdec --output help`.
- That aside I tried to keep carried-over short options unchanged.
- A side effect of the reorganization is that all options - *including frequencies* - can be provided in any order. In other words, you can specify options after frequencies and even *between* frequencies if you so fancy.

I tried to revamp the help text (shown with `acarsdec -h`) to make everything as clear as possible, as well as correctly handling the cases where not all optional libraries were available at build time.

Finally the things I didn't touch (beyond code cleanup) at this stage are:
- output formats syntax (for obvious backward compatibility reasons): I've been testing "acarsdec 4.0" feeding JSON to airframes.io with no problem.
- Airspy SDR backend, which does things rather differently than the other 3 (beyond the filter tweak, that is) for reasons I'm not completely clear about: I'd like to make changes but I'd like to get some feedback from the author before I do (e.g. why is it operating entirely on real numbers instead of complex ones, which as far as I can tell the airspy driver can provide?). So the performance of this backend should be essentially unchanged as it doesn't use the new optimized common code.
- MSK demodulator
- ACARS decoder

The latter two I might revisit later as I see some room for improvement (I'd like to add a noisefloor measurement among other things), but I thought that a 100-commit PR was a nice round number for a start 😅

I have tried to add some meaningful documentation where applicable, I'm by no means an SDR expert so I hope I didn't write any blunder. Should probably still be better than nothing 😊

Sucessfully built: everything except the sdrplay backend (I don't have the required dependencies)

Tested:
- rtlsdr input
- soapysdr input (both rtlsdr and network modules)
- sndfile input
- file output
- udp output
- mqtt output
- all output formats
- statsd reporting

Needs testing:
- sdrplay input
- airspy input
- ALSA input

(though I shouldn't have brought any functional changes to the latter two)

Tests performed on macOS x86_64, Linux x86_64 and armhf.

I think that sums it up. I'm happy to help maintain acarsdec in the future, if that can help 😊

I hope all of this is useful.

Commit:
Add documentation for recent patches

https://github.com/TLeconte/acarsdec/pull/106

Commit:
Allow 2.5MHz bandwidth on R2 use without filter optimization if needed.

https://github.com/TLeconte/acarsdec/pull/107

murrayma committed:
Use sendto(2) for udp and retry dns on temporary failure.

https://github.com/TLeconte/acarsdec/pull/111

murrayma committed:
Prevent clobbering the destination hostname and better error reporting

https://github.com/TLeconte/acarsdec/pull/111

Commit:
added RTLSDR bias tee support

https://github.com/TLeconte/acarsdec/pull/116

f00b4r0 committed:
netout: some cleanup & improvements

- avoid special-casing IPv6 (AF_UNSPEC already handles it)
- avoid keeping multiple copies of the Rawaddr parameter
- avoid redundant (and useless) error checks
- avoid redundant checks in Netout*()
- avoid running strlen() over buffer when snprintf() is guaranteed to
  return the buffer size
- avoid duplicating the prepared JSON buffer, send it natively instead
  (tested working with airframes.io)
- use static buffers for Netoutpp() and Netoutsv() to reduce stack pressure

f00b4r0 committed:
CMakeList: fixes

Fixes: https://github.com/TLeconte/acarsdec/issues/112

f00b4r0 committed:
quick ugly hack to allow building multiple backends

- move common global variables out of #define scope
- exit with an error if more than one input is selected
- reformat the usage help text
- rename rtlMult -> rateMult (in line with soapy)
- adjust rtl and soapy gain routines to work with floats

cmake -Drtl=ON -Dsoapy=ON -Dairspy=ON works and builds

f00b4r0 committed:
rtl: always look for serial if device id is 8-char long

This fixes the following problem:

Found 3 device(s):
  0:  Realtek, RTL2838UHIDIR, SN: 00000003
  1:  Realtek, RTL2838UHIDIR, SN: 00000002
  2:  Realtek, RTL2838UHIDIR, SN: 00000001

`acarsdec -r 00000002` would select the device with serial 00000001
`acarsdec -r 00000001` would select the device with serial 00000002

This patch disambiguates explicit serials by skipping matching device ID
number when the passed argument is exactly 8 characters long, like a
fully qualified serial.

This patch is designed to be as short a diff as possible while preserving
existing behaviour for all but 8-character long arguments.

f00b4r0 committed:
Revert "add watchDog to detect hanging rtl-sdr"

This reverts commit 22f0c810c8bf1ce5e2ee56d062a1eb8a2e1878b6.

Does this solve an actual problem? dumpvdl2 doesn't have similar code.
This runs an extraneous thread and uses nonportable pthread API.
-Drtl=ON now builds on macOS

f00b4r0 committed:
get rid of pthread in soapy

Does this solve an actual problem? Neither dumpvdl2 nor common examples
have similar code. This runs an extraneous thread and uses nonportable
pthread API. -Dsoapy=ON now builds on macOS

Additionally:
- add some error checking
- retry on SOAPY_SDR_OVERFLOW (previous code wouldn't work either without it)
- gracefully retry on zero-reads
- don't free in buffer while read() may still be in-flight
- remove unnecessary global variables

PS: this needs so much more work :/

f00b4r0 committed:
Remove some unused includes and global vars

f00b4r0 committed:
code cleanup: regroup all runtime global vars under a single namespace

- No functional change.
- Some unused vars have been removed.

f00b4r0 committed:
code cleanup: deduplicate frequencies parsing

- No functional change.
- Some unused vars have been removed.

f00b4r0 committed:
gitignore: add build/ folder

f00b4r0 committed:
code cleanup: deduplicate chooseFc()

- Simplify the logic (use the median value of min and max frequency)
- Switch all Fc/minFc/maxFc references to unsigned int

This also "fixes" the following warning:
warning: taking the absolute value of unsigned type 'unsigned int' has no effect [-Wabsolute-value]

f00b4r0 committed:
code cleanup

First pass at cleaning the mess, using Linux kernel clang-format rules
(without ColumnLimit) and a handful of manual adjustements.

No code change.

f00b4r0 committed:
acarsdec: cleanup usage()

don't explicitly zero R initializers (zeroing is implicit)

f00b4r0 committed:
rename R.channel -> R.channels

f00b4r0 committed:
main(): use errx() where applicable

f00b4r0 committed:
make number of channels dynamic

allows decoding an arbitrary number of channels, limited only by CPU
and memory.

channels.Fr is now unsigned int, adjust casts accordingly until further
cleanup.

f00b4r0 committed:
code cleanup: s/sizeof(type)/sizeof(var)/

Remove malloc() casts.
No code change.

f00b4r0 committed:
code cleanup: use system-provided cJSON

find it through pkg-config since we're already using that for libacars
and make it a hard dependency for now.

f00b4r0 committed:
code cleanup: improve CMakeLists.txt

We no longer need the various -D<radio>=On since we can build them all
simultaneously. Instead, print an informative message when the
corresponding library is or isn't found.

f00b4r0 committed:
code cleanup: make libcJSON optional

also make MQTT support depend on it.

f00b4r0 committed:
netout: add minimal sanity checks

f00b4r0 committed:
rtl: fix error message

f00b4r0 committed:
treewide: first step for making outputs independent

This commit enables the use of multiple concurrent outputs, using the
following syntax:

acarsdec --output format:destination:parameters [--output ...]

Where "format" is one of:
- "oneline" for single line text decoding
- "full" for full text decoding
- "monitor" for live decoding
- "pp" for PlanePlotter
- "native" for Acarsdec native format
With CJSON support enabled:
- "json" for JSON output
- "routejson" for flight route output in JSON format

and "destination" is one of:
- "file" for file (including stdout) output
- "udp" for network output over UDP
With MQTT support enabled:
- "mqtt" for MQTT output

Not all combinations of format and destination are valid, acarsdec will
complain if an invalid combination is chosen.

params are described in respective source file (TODO: update usage())

Fixes: https://github.com/TLeconte/acarsdec/issues/76

f00b4r0 committed:
fileout: make rotation a per-output parameter

f00b4r0 committed:
treewide: code reordering

No functional change

f00b4r0 committed:
fileout: fix logfile suffixing

f00b4r0 committed:
allow setting center freq for all but Airspy SDRs

f00b4r0 committed:
treewide: more disentangling of acarsdec.h

f00b4r0 committed:
output: avoid unnecessary zeroing

calloc() already zeroes out allocated memory

f00b4r0 committed:
alsa.c remove WITH_ALSA fencing

(NB: is this even working? It's not being built ever)

f00b4r0 committed:
acarsdec: move SDRPLAY to "-S"

It conflicted with airspy.

f00b4r0 committed:
Update README and usage()

f00b4r0 committed:
treewide: remove unncessary fencing, fix build with DEBUG=1 and SNDFILE=NO

f00b4r0 committed:
Implement --output help

f00b4r0 committed:
code cleanup: air/rtl: rename "ch->wf" to "ch->oscillator"

This removes an extra variable from channel_t

Also remove unused pthread_t th and associated include (add include to
acars.c where pthreads are used).

f00b4r0 committed:
code cleanup: {m,c}alloc() error checking and reporting

Add missing checks, report via perror() for consistency.

f00b4r0 committed:
code cleanup: soapy: remove redundant frequency range check

Already performed in find_centerfreq()

Removing the check also allows manual setting ("-c") to override warning

f00b4r0 committed:
code deduplication: channels_init_sdr()

Airspy does its thing differently, leave it be for now.

For Soapy, this moves an immediate FPdiv out of the sample processing
loop and should improve CPU load.

f00b4r0 committed:
code cleanup: make unsigned variables actually unsigned

f00b4r0 committed:
CMakeLists: enable "-Wall" and fix warnings

f00b4r0 committed:
code cleanup: remove dead code

f00b4r0 committed:
check for libasound to enable ALSA support

f00b4r0 committed:
find_centerfreq(): use multiplier instead of samplerate

This aligns the API with channels_init_sdr() and allows us to remove a
few unused variables. It also makes the spread check clearer.

f00b4r0 committed:
Add behavior documentation in lib/rtl, rewrite rtl callback

This provides an explanation to what's going on and should help aligning
and deduplicating other dongle code. It also rewrites rtl in_callback()
in a way that is more consistent.

This results in a slight reduction in CPU usage.

f00b4r0 committed:
soapy: reverse loop order in runSoapySample()

do not repeatedly traverse the entire buffer once for each channel.
This reduces CPU usage by 25%

f00b4r0 committed:
acarsdec: fix signal handling

Fixes: cec01f571

f00b4r0 committed:
soapy: rewrite runSoapySample() to use local variables

This removes the need to store redundant (same value) variables in the
channel_t structure and should improve CPU usage by reducing the number
of dereferences and improving cache locality.

f00b4r0 committed:
rtl: fix autogain logic

Autogain would be enabled for any gain setting below 10, instead of -10.

f00b4r0 committed:
soapy: fix runSoapyClose()

Correctly order deconstructors and avoid NULL dereference crash on exit.

f00b4r0 committed:
soapy: reorder code and fail on fatal errors

f00b4r0 committed:
soapy: switch to native CF32 and use smaller input buffer

This lets Soapy handle conversion and scaling and simplifies the sample
processing routine. We also don't need a huge buffer since we are doing
incremental processing and Soapy will typically limit the actual used
size when used in network mode.

As an added bonus it also fixes the reported message level which is now
in line with native rtlsdr.

f00b4r0 committed:
make R.inmode an enum

f00b4r0 committed:
acarsdec: move input init out of getopt()

init after all opts have been processed. This also fixes default gain.

f00b4r0 committed:
code cleanup: rework input initialization

parse frequencies in a single place, avoid processing the argument array
within input initializers. This also makes it possible to have option
arguments *after* frequencies, e.g.:

  acarsdec --output full:file -r 0 131.550 130.825 130.850 -g 40

now works as intended

f00b4r0 committed:
soapy: code cleanup: move antenna setup to initSoapy()

f00b4r0 committed:
let's stop pretending this is still acarsdec 3.X ;)

f00b4r0 committed:
sdrplay: don't use channel-local extra variables

This is an ugly fix, the correct way would be to adopt the same method
as soapy.c, but I can't build this and it seems it's using a deprecated
API anyway so I have no interest in doing better :P

f00b4r0 committed:
update usage() and README

f00b4r0 committed:
lib: implement channels_mix_phasors()

Rationalize phasor processing across SDRs by providing one routine to
rule them all. Convert rtl and soapy.

For soapy, the cumulated perf boost vs 3.7 is results in CPU usage
divided by ~4: there is no significant performance degradation in using
soapy driver=rtlsdr vs using native rtl code.

rtl can now deal with partial buffers, if that's ever needed.

f00b4r0 committed:
code cleanup: air: don't pollute global namespace

also add a REVIEW note: it looks like things could be simpler

f00b4r0 committed:
air: remove usage of channel-local D accumulator

f00b4r0 committed:
channel_t: remove two now-unused variables

f00b4r0 committed:
Initial statsd implementation

f00b4r0 committed:
improve struct packing

Checked on 32 and 64bit archs

f00b4r0 committed:
acars: blk_thread() access blk->chn only once

f00b4r0 committed:
acarsdec: fix errx() strings

f00b4r0 committed:
acarsdec: hook statsd init

f00b4r0 committed:
acars: add some statsd stats

Add statsd hooks into the output thread, so as not to slow down the
input processing.

f00b4r0 committed:
sdrplay: use channels_mix_phasors(), fix scaling [UNTESTED]

Adjust what looked like a wrong scaling factor for signed 16bit integers
Switch to channels_mix_phasors() and simplify the callback loop.

This is completely untested.

f00b4r0 committed:
code cleanup: add msk.h

Now that demodMSK() isn't used everywhere anymore, we can cleanup the
namespace a bit.

f00b4r0 committed:
acarsdec: try to improve usage() again

f00b4r0 committed:
CMakeList: don't forcefully set -march=

Instead use the compiler default and allow the user to override as
needed via CMAKE_C_FLAGS.

This will make cross-compiling easier, make a potential package
maintainer's life easier and yield default builds that will always work.

f00b4r0 committed:
update usage() and README about statsd

also mention paho-mqtt3a optional dep

f00b4r0 committed:
Add LICENSE (GPL-2.0)

f00b4r0 committed:
acarsdec: use long opts for inputs

option names loosely based on dumpvdl2, for consistency.

This disambiguates, makes room for more input options should there ever
be more, and frees '-a' for antenna selection.

Rewrite usage() and README one more time

f00b4r0 committed:
rtl: one less static variable

f00b4r0 committed:
move parse_freqs() to acarsdec.c

It's only used there.

f00b4r0 committed:
treewide: error message consistency

f00b4r0 committed:
alsa: fix missing channels malloc and error check

f00b4r0 committed:
dos2unix soapy.c

f00b4r0 committed:
remove single-use typedef sample_t

f00b4r0 committed:
lib,rtl,soapy: use un/likely()

Tell the compiler what we expect in cpu-bound code

f00b4r0 committed:
Update README

f00b4r0 committed:
remove #define _GNU_SOURCE

Nothing seems to require this.

f00b4r0 committed:
code cleanup: use a single dm_buffer size

f00b4r0 committed:
treewide: do not randomly exit()

f00b4r0 committed:
soapy: fix signal handling

f00b4r0 committed:
soapy: tidyup initSoapy()

f00b4r0 committed:
rtl: tidyup initRtl()

f00b4r0 committed:
runtime_t: reorder for cacheline boundary optimization

This may cause some holes: fairly harmless for a single-instance struct

f00b4r0 committed:
soapy: runSoapySample(): cache R.rateMult

f00b4r0 committed:
treewide: use a global running flag for exit control

f00b4r0 committed:
soundfile: error check order

f00b4r0 committed:
acarsdec 4.0

f00b4r0 committed:
treewide: define and use vprerr() for verbose messages

f00b4r0 committed:
acars: sync on entire PREKEY

Instead of just syncing on two (possibly inverted) SYN bytes, which is
rather prone to error, sync on the entire PREKEY (which is designed for
this exact purpose). Also perform phase inversion during PREKEY.

PREKEY will be currently accepted after 12 or more 0xFF sync pulses, to
be adjusted once the MSK PLL lock acquisition speed is reviewed.

The net result is that decodeAcars() now syncs much more reliably and
always requests entire bytes from demodMSK() (with only one exception),
instead of spending most of the time processing bit by bit, waiting for
a sync signal. This also improves CPU efficiency.

f00b4r0 committed:
msk: rework level estimator

Use an exponential moving average over the last 16*8 bits.
This simplifies signal level handling and gives results similar to the
previous method, see comparison below.

Old:
./acarsdec -f ../test.wav -o 1
#2 (L:-23.0 E:0)   PH-BXR KL1681 E 5V S53A
#2 (L:-19.1 E:0)   LN-DYY DY083J E Q0 S47A
#4 (L:-17.4 E:0)   LN-DYY DY083J 2 Q0 S46A
#1 (L: -8.5 E:0)   F-GTAE AF7728 G H1 D65C 00000/V206,05,124,183,02,00,00000/V3XX,XX,XXX,XXX,XXXX/V4XX
#1 (L:-18.6 E:0)   LN-DYY        x _d
#3 (L: -4.6 E:0)   G-DBCK BA031T 2 _d S64A
#3 (L: -4.3 E:0)   G-DBCK BA031T E Q0 S63A

New:
./acarsdec --sndfile ../test.wav --output oneline:file
#2 (L:-23.3 E:0)   PH-BXR KL1681 E 5V S53A
#2 (L:-19.3 E:0)   LN-DYY DY083J E Q0 S47A
#4 (L:-17.7 E:0)   LN-DYY DY083J 2 Q0 S46A
#1 (L: -8.4 E:0)   F-GTAE AF7728 G H1 D65C 00000/V206,05,124,183,02,00,00000/V3XX,XX,XXX,XXX,XXXX/V4XX
#1 (L:-18.6 E:0)   LN-DYY        x _d
#3 (L: -4.8 E:0)   G-DBCK BA031T 2 _d S64A
#3 (L: -4.5 E:0)   G-DBCK BA031T E Q0 S63A

f00b4r0 committed:
acars: show level in verbose sync message

This helps diagnose failed syncs due to too low SNR

f00b4r0 committed:
acars: make init more robust

Don't depend on channel number (and channel init order)
Use statically initialized values since we are using the defaults anyway

f00b4r0 committed:
MskLvl does not require double precision

f00b4r0 committed:
acars: use __builtin_parity() when available

On modern hardware this will use a single insn and save us a table
lookup. And on other platforms, the compiler can still be very efficient

f00b4r0 committed:
acars: rewrite update_crc() macro to unhide assignment

and be explicit about crc size

f00b4r0 committed:
acars: move put block to END state

For extra pedantry, report missing DEL in verbose mode

f00b4r0 committed:
acars: un/likely() - reorder some checks for speed

f00b4r0 committed:
acars: simplify blk chain list

There is no need to keep track of the list end since this is a singly
linked list anyway.

f00b4r0 committed:
acars avoid redundant checks, avoid reporting crc errors on parity

f00b4r0 committed:
acars: abort early if parity check fails again

f00b4r0 committed:
acars: blk_thread(): regroup error path

f00b4r0 committed:
acars: don't check parity during reception

this is completely redundant with the check in blk_thread() and adds
extra processing time in the sample processing path.

f00b4r0 committed:
acars: make parity removal loop vectorizable

f00b4r0 committed:
output: use faster mem*() instead of str*()

This is faster, especially considering that the source will typically be
all-but-last non-nulls.

f00b4r0 committed:
outputmsg(): remove unnecessary assignments

msg is zeroed on initialization

f00b4r0 committed:
output: improve downlink msg parsing logic

Use memcpy() instead of for loops and entirely skip parsing if text_len
is too short. Avoids redundant tests.

f00b4r0 committed:
label: use switch and immediate return

This speeds up execution by not trying each if() in sequence and makes
the function operation clearer. Using immediate returns instead of a
function pointer and single return point allows aggressive auto-inlining

f00b4r0 committed:
addFlight(): don't use now unnecessary intermediary variable

f00b4r0 committed:
fmt_msg(): fix 'Wheels off Tme' typo

f00b4r0 committed:
DecodeLabel(): do not zero oooi argument

f00b4r0 committed:
output: fix out of bound accesses, avoid duplicating txt

This reduces memory pressure and cache trashing and speeds up execution.
outputmsg() no longers allocates memory unnecessarily and operates on
the data passed on from acars blk_thread(), which it only modifies once
(nul-termination at suffix position - necessary as cJSON doesn't have
primitives to limit string length).

f00b4r0 committed:
acarsmsg_t: remove unneeded member

'bs' is only used once locally in outputmsg()

f00b4r0 committed:
move acarsmsg_t from label.h to acarsdec.h

use explicit size relationships for null-terminated copies

f00b4r0 committed:
output: remove unnecessary #ifdefs

f00b4r0 committed:
acarsmsg_t: remove unneeded member

'msn_seq' is only used once locally in outputmsg()

f00b4r0 committed:
acarsmsg_t: pack

Move one member outside of #ifdef since it's a standard type and this
makes packing deterministic.

f00b4r0 committed:
soundfile: make inner sample read loop vectorizable

use sf_readf_float() and close input file on exit

f00b4r0 committed:
treewide: add missing catchall error for token parameters

f00b4r0 committed:
netout: add missing parameter check

f00b4r0 committed:
treewide: deduplicate token parameter parsing

f00b4r0 committed:
parse_params(): allow processing of lone tokens

f00b4r0 committed:
soundfile: add support for raw audio input

This partly addresses https://github.com/TLeconte/acarsdec/issues/109
until a proper raw IQ input is added, and replaces
https://github.com/TLeconte/acarsdec/pull/110 as a more generic
implementation

use '--sndfile <file>' as previously for all containers supported by
libsndfile.
use '--sndfile <file>,subtype=<N>' for raw audio processing with <N>
being a supported subtype value as specified in sndfile.h. Optionally,
'channels=<M>,endian=<end>' can be added to decode <M> channels (default
1 if not specified) using <end> endianness, where <end> is one of 'big',
'little' or 'cpu' (default 'cpu' if not specified).

f00b4r0 committed:
soundfile: support sample rates which are multiple of INTRATE

use '-m' to change the multiplier from the default '1'

f00b4r0 committed:
update README

f00b4r0 committed:
msk: use ternary op to improve compiler optimization

for the if (ch->MskS & 1) block, using a ternary op enables the compiler
to efficiently optimize the sign adjustment and real/imag part assignment
using (on x86) a single branch instead of 5.
for the if (ch->MskS & 2) block, using a ternary inside the function
call reduces code size by having only one inlined copy and allows further
integration with the results of the previous if block.

this also makes the code slightly more compact/readable

f00b4r0 committed:
treewide: rework usage of debug output WAV

dump all channels in a more efficient manner than doing it sample by
sample

f00b4r0 committed:
msk: do not divide power by 4

As dm_buffer from SDR input only contains positive values (cabsf()),
i.e. half of the power scale (just like the matched filter), this
normalization skews the actual dB scale readout and prevents reporting
clipping (i.e. overloads).

This change has been matched with a readout of an audio dump of a live
SDR capture normalized in an audio processing software: the reported
value of the normalized file is '-0.0' as expected, vs '-6.0' with the
previous code. The lowest reported values for SDR input are now also
more in line with the expected SNR.

In order to keep soundfile input values consistent with this change, the
audio samples are "normalized" to half scale. That means that audio
files generated through DEBUG will appear 6dB quieter than the original
signal (i.e. the readback will report the same value as was reported for
SDR input *before* this patch). The rationale is to show "true" values
for "standard" audio input.

f00b4r0 committed:
acars: report post sync signal level

This measures signal level averaged mostly over the PRESYNC key, which
should typically be the "loudest" part.

f00b4r0 committed:
acars: report loud messages (above 0dB)

f00b4r0 committed:
msk/acars: add support for noise floor estimator

This introduces a basic, low cpu-cost NF estimator which works by using
an exponential moving average over the magnitude of the received signal
updated only outside of ACARS message processing.

The computational overhead of this extra estimator is virtually nil, as
the signal power estimator has been reworked to use the same input as
the noise floor estimator.

The known limitation is that this estimate will be skewed in case of a
large burst of ACARS messages as the average will no longer converge
towards the noisefloor. To be improved on later.

Fixes: https://github.com/TLeconte/acarsdec/issues/84

f00b4r0 committed:
msk: rename some defines for clarity, document some members

use a ceiling macro to compute correct BITLEN regardless of INTRATE

f00b4r0 committed:
msk: scale matched filter to BITLEN, adjust reported level

This commit enables correct support of arbitrary INTRATE values (as long
as they respect Nyquist-Shannon)

This corrects d721086

f00b4r0 committed:
output: silence a GCC warning

GCC is apparently not as smart as clang to identify uninitialized uses

Fixes: d6b48ae

f00b4r0 committed:
INTRATE = 12kHz

Integer multiple of MSK shift freq, this should result in less bit phase
drift (cf BITLEN). This also enables "native" 48kHz audio processing.

f00b4r0 committed:
acars: don't use hardcoded numbers for boundary checks

f00b4r0 committed:
acars: use an union to access fixed-size members of TXT segment

f00b4r0 committed:
rtl: set receiver bandwidth

This narrows the IF filter band to improve rejection of nearby unwanted
signals.

f00b4r0 committed:
msk: oscillator phase does not require double precision

f00b4r0 committed:
acars: refactor decodeAcars()

f00b4r0 committed:
acars: use stdint types

f00b4r0 committed:
acarsdec 4.1

fredclausen commented:
Thank you so much for this. I maintain a [docker container](https://github.com/sdr-enthusiasts/docker-acarsdec) that originally was running TLeconte's acarsdec, and I've switched to using your version 4.0/4.1 for my users. It works wonderfully, and I appreciate your work on this.

f00b4r0 commented:
> Thank you so much for this. I maintain a [docker container](https://github.com/sdr-enthusiasts/docker-acarsdec) that originally was running TLeconte's acarsdec, and I've switched to using your version 4.0/4.1 for my users. It works wonderfully, and I appreciate your work on this.

Thanks for the feedback, much appreciated. I have a few more improvements brewing but have left them on the backburner especially considering there was zero feedback here. This gives new motivation to complete that work 👍

f00b4r0 commented:
In case the brave and adventurous are interested, testing of this branch is most welcome:

https://github.com/f00b4r0/acarsdec/commits/demod/

There are a few changes to the decoder for which the most user-side visible impact is a change in reported signal levels (which should be closer to "reality") and the addition of a (very basic and not very accurate) noise floor estimate (shown in plain text output: `L` now shows "signal/noise"). The internal samplerate has also been switched to 12kHz (from 12.5) meaning that processing e.g. 48kHz "native" audio sample rate is now possible without intermediary resampling.

NB: I'm running that code myself but _don't assume any/all of these changes are release-worthy_ (i.e.: will be merged into master), this is **work in progress** left as it was when I put things on hold. FWIW I'm also successfully using a much lower multiplier (`-m`) than current default (in fact, the minimum allowed by my dongle to grab all ACARS channels I'm monitoring) which roughly translates into halved CPU load, suggesting the current default may not be ideal: I might revisit this.

I still want to improve the MSK routines (need to better understand the PLL lock speed and a couple tricks used there) before I move forward. The filter also needs improvement as I'm occasionally seeing messages spillover across adjacent channels (this was already the case with stock acarsdec so probably not related to my changes).

HTH

f00b4r0 committed:
soapy: set receiver bandwidth

See d389b4

wiedehopf commented:
demod branch seems to work fine.

master branch of your fork is in use for https://github.com/sdr-enthusiasts/docker-dumpvdl2 which in turn is used by adsb.im

Might be useful to open issues on your fork just because.
Though so far it just works very nicely.

f00b4r0 commented:
> demod branch seems to work fine.

Thanks for the feedback. I'll merge into master soon I guess. I'd still want to improve the matched filter (there's too much crosstalk between adjacent channels on loud signals - unrelated to my changes) but didn't find the time to wrap my head around it yet. I need to educate myself a little more about SDR signal processing tricks.

> master branch of your fork is in use for https://github.com/sdr-enthusiasts/docker-dumpvdl2 which in turn is used by adsb.im

Cool! I suppose you meant https://github.com/sdr-enthusiasts/docker-acarsdec (even though the README mentions some other fork) ;)

> Might be useful to open issues on your fork just because.

Have to admit being reluctant to do that as I'm not overly keen on dealing with the typical GH issue onslaught. ;P

At least PRs tend to be more productive. But I guess I'll open the issues now that this is getting more audience.

> Though so far it just works very nicely.

Great. Thanks again for the feedback, much appreciated.

f00b4r0 committed:
soapy: better default sample rate multiplier

Default to 100 (1.2MS/s) which should be plenty enough for most
situations while reducing the CPU load from the previous default, until
proper automation is implemented (requires parsing the target device
capabilities).

f00b4r0 committed:
rtl: automatic sample rate multiplier

Compute the minimum usable sample rate multiplier, flooring at 80
(960kS/s) to account for the RTL sampling rate hole.

f00b4r0 committed:
CMakeList: fix compilation with modern CMake

f00b4r0 committed:
README update

wiedehopf commented:
--rtl-mult auto
Would that be simple enough to implement?
Otherwise i might just try to implement it in the container.
As in in the scripting that starts up acarsdec.

i've tested 80 and lower is probably not terribly useful as you approach the limits of the rtl-sdr anyhow.

f00b4r0 commented:
No worries, I'll merge and will open issues. We should take these conversations to the right repo.

f00b4r0 committed:
acarsdec 4.2

f00b4r0 committed:
test.wav: resample to 12kHz

f00b4r0 committed:
treewide: let backends handle their own default configs

this will be necessary to use SoapySDRDevice_getSampleRateRange() for
Soapy automagic multiplier without adding more #ifdefs

f00b4r0 committed:
lib: purge dead code

f00b4r0 committed:
sdrplay: fix static multiplier

SDRPlay supports arbitrary sample rate between 2 and 10 MSps.
With the change of INTRATE from 12500 to 12000, the hardcoded value
would fall out of this range. This fixes it.

Fixes: 6fc3de1

f00b4r0 committed:
soapy: automatic sample rate multiplier

This selects the minimum suitable sample rate based on hardware
capabilities and chosen frequencies.

It also reports if a manually selected rate multiplier is out of range
for the hardware.

Fixes: 5ff20d3
Fixes: #1

f00b4r0 committed:
Update soapy usage and README

f00b4r0 committed:
air,sdrplay: remove dangling refs to compute_thread

f00b4r0 committed:
rtl: fix RTLMULTMAX

f00b4r0 committed:
sdrplay: remove unnecessary global variable

f00b4r0 committed:
make it clear that airspy and sdrplay are unmaintained

f00b4r0 committed:
get version from git

f00b4r0 committed:
acarsdec 4.3

wiedehopf committed:
fix compile failure with CMake < 3.26

Extraneous '-D' is only removed by CMake >= 3.26

wiedehopf committed:
add missing include

output.c:45:1: note: ‘offsetof’ is defined in header
‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?

wiedehopf committed:
airspy samplerate selection: use lowest available

use this rather than rely on the ordering of samplerates

wiedehopf committed:
airspy: simplify code by using CF32 samples

it would seem the previous code wasn't working
after this commit i can get decodes

wiedehopf committed:
fixup soapy sdr sample rate selection

airspy returns the sample rates in reverse order
sort them so the sample rate selection algorithm is fixed

wiedehopf committed:
airspy: add biastee support

f00b4r0 committed:
Update README

f00b4r0 committed:
soapy: switch to mergesort()

mergesort() is stable and more efficient when the data is already
(reverse) ordered

f00b4r0 committed:
airspy: remove now unused global variable

Fixes: 4cf0abe

f00b4r0 committed:
airspy: code cleanup, no functional change

f00b4r0 committed:
airspy: -m and -p are ignored

f00b4r0 committed:
airspy: cleanup init routine

f00b4r0 committed:
lib: document channels_init_sdr()

f00b4r0 committed:
airspy: airspy_exit() is deprecated

f00b4r0 committed:
airspy: add error message prefix

f00b4r0 committed:
soapy: do not use static type for sizeof()

f00b4r0 committed:
airspy: use find_centerfreq()

the fancy filter setup apparently only works for real mode and besides
only at 5MSps; either case we're not using anymore.

Now airspy will honour '-c', like all other SDR backends

f00b4r0 committed:
update README and usage()

reorder options help alphabetically

f00b4r0 committed:
Revert "soapy: switch to mergesort()"

mergesort() is a BSD thing. It's not worth pulling in libbsd for this.

Fixes: 791abe5

f00b4r0 committed:
sdrplay: move default values to sdrplay.c

Fix the GRdB default value to match usage and README

TLeconte commented:
Many thanks to Thibaut for continuing acarsdec dev and for acarsdec 4.x

I will archive this repository and point to https://github.com/f00b4r0/acarsdec/ for futur versions.
